grudge/dt_utils.py

```diff
@@ -46,21 +58,65 @@
 
 
 @memoize_on_first_arg
-def dt_non_geometric_factor(dcoll: DiscretizationCollection, dd=None) -> float:
-    r"""Computes the non-geometric scale factor following [Hesthaven_2008]_,
-    section 6.4:
+def characteristic_lengthscales(dcoll: DiscretizationCollection) -> DOFArray:
```

Comment:
At the dev meeting this morning, @MTCam pointed out that, usually, the very next thing folks will do with the result from this is some type of arithmetic, so it's awkward if they have to thaw it. As a result, this should probably take an array context and return a thawed result.

As far as the memoization is concerned, you can simply take this function, rename it to say `_characteristic_lengthscales_inner`, and provide a "thawing wrapper".

Reply:
Done here: https://github.com/inducer/grudge/pull/121/commits/ec7ca8345bfacd50be6d24c54e1bfc15f29fcff6